### PR TITLE
pkg/trace/api: add live debugger proxy endpoint

### PR DESCRIPTION
**Pull request by ojung**
### What does this PR do?

Add an endpoint to forward requests the logs intake.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Describe how to test your changes

Write here in detail how you have tested your changes
and instructions on how this should be tested in QA.

### Checklist

- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

### CODE REVIEW

File: pkg/trace/api/api.go [hunk 1]
```diff
@@ -762,3 +763,34 @@ func getMediaType(req *http.Request) string {
 	}
 	return mt
 }
+
+// getDirector returns a director for a httputil.ReverseProxy. It enriches requests with headers setting custom
+// container and additional tags and overwrites the user agent header.
+func getHeaderDecoratorDirector(tags string, metricSuffix string) func(req *http.Request) {
```
Comment (gbbr) [inline pkg/trace/api/api.go:769]: In general, in Go, we like to avoid the `get*` prefix (see https://golang.org/doc/effective_go#Getters - granted this isn't a getter). Of course this is not a hard rule and exceptions can be made, but I'd prefer we do it here too.
```suggestion
func headerDecoratorDirector(tags string, metricSuffix string) func(req *http.Request) {
```

File: pkg/trace/api/api.go [hunk 2]
```diff
@@ -762,3 +763,34 @@ func getMediaType(req *http.Request) string {
...
+			// that net/http gives it: Go-http-client/1.1
+			// See https://codereview.appspot.com/7532043
+			req.Header.Set("User-Agent", "")
+		}
+		containerID := req.Header.Get(headerContainerID)
+		if ctags := getContainerTags(containerID); ctags != "" {
+			req.Header.Set("X-Datadog-Container-Tags", ctags)
+		}
+		req.Header.Set("X-Datadog-Additional-Tags", tags)
+		_ = metrics.Count("datadog.trace_agent."+metricSuffix, 1, nil, 1)
```
Comment (gbbr) [inline pkg/trace/api/api.go:783]:
```suggestion
		metrics.Count("datadog.trace_agent."+metricSuffix, 1, nil, 1)
```
It's fine to ignore the error. No need to assign to `_`. If you did that because of your editor, perhaps there is some setting somewhere.

File: pkg/trace/api/api.go [hunk 3]
```diff
@@ -762,3 +763,34 @@ func getMediaType(req *http.Request) string {
...
+	}
+}
+
+// getOrchestrator returns the fargate orchestrator. Should only be called if the agent is running in ECS or EKS.
+func getOrchestrator() string {
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	orch := fargate.GetOrchestrator(ctx)
+	cancel()
+	if err := ctx.Err(); err != nil && err != context.Canceled {
+		_ = log.Warnf("Failed to get Fargate orchestrator. This may cause issues with your snapshots: %v", err)
```
Comment (gbbr) [inline pkg/trace/api/api.go:793]: Same here.

File: pkg/trace/api/api.go [hunk 4]
```diff
@@ -762,3 +763,34 @@ func getMediaType(req *http.Request) string {
...
+		if ctags := getContainerTags(containerID); ctags != "" {
+			req.Header.Set("X-Datadog-Container-Tags", ctags)
+		}
+		req.Header.Set("X-Datadog-Additional-Tags", tags)
+		_ = metrics.Count("datadog.trace_agent."+metricSuffix, 1, nil, 1)
+	}
+}
+
+// getOrchestrator returns the fargate orchestrator. Should only be called if the agent is running in ECS or EKS.
+func getOrchestrator() string {
```
Comment (gbbr) [inline pkg/trace/api/api.go:788]: Why not call it `fargateOrchestratorTag` (I do prefer names to be as short as possible but I think we need to be specific here) and return the value `"orhcestrator:"+strings.ToLower(orch)` since this is what happens in both places. That would then be even more de-duplication.

File: pkg/trace/api/api.go [hunk 5]
```diff
@@ -762,3 +763,34 @@ func getMediaType(req *http.Request) string {
...
+		containerID := req.Header.Get(headerContainerID)
+		if ctags := getContainerTags(containerID); ctags != "" {
+			req.Header.Set("X-Datadog-Container-Tags", ctags)
+		}
+		req.Header.Set("X-Datadog-Additional-Tags", tags)
+		_ = metrics.Count("datadog.trace_agent."+metricSuffix, 1, nil, 1)
+	}
+}
+
+// getOrchestrator returns the fargate orchestrator. Should only be called if the agent is running in ECS or EKS.
```
Comment (gbbr) [inline pkg/trace/api/api.go:787]: Well, actually, this "should only be called if ..." won't really guarantee that people won't do it. It should be ok to call this regardless and have it return `""`. The only way to achieve that is more work:
* Remove `(*AgentConfig).IsFargate` field
* Use `(pkg/util/fargate).IsFargateInstance` but create a global variable cache of the first response which stores the value to avoid executing the current function body which is an HTTP request on each call.

Or, if you don't want to do the above, then simply ignore the above along with my other comment about returning the full tag and accept that this function may be called even outside a Fargate instance and document that it will return `(pkg/util/fargate).Unknown` and handle that case in code where this is called.

Lastly, as a note, `fargate.GetOrchestrator` will involve an HTTP call too, every time. We could avoid this and only do it once at startup by replacing what is now `(*AgentConfig).IsFargate` with `(*AgentConfig).FargateOrchestrator` and just know that the value `Unknown` means we are not in Fargate. That way, this config field can account both for telling us whether we are in Fargate (like today) and also the orchestrator.

File: pkg/trace/api/debugger.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+type roundTripper struct {
+	url *url.URL
+	key string
+	rt  http.RoundTripper
+}
+
+func (r *roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.Header.Set("DD-API-KEY", r.key)
+	req.URL = r.url
+	req.Host = r.url.Host
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:79]: In reality, if we are to be honest with one another, this is only here because you wanted to de-duplicate some code. If you wouldn't have de-duplicated the code, the API key could be set in the Director and this whole structure could be dropped.

The only way to make use of this is:
* If you need other logic around the roundtripper (such as additional endpoints like profiling), or,
* You want to add metrics based on how this goes (highly recommended), for example: `datadog.trace_agent.debugger.proxy_request`, `datadog.trace_agent.debugger.proxy_error`, etc.

**Review by gbbr — COMMENTED** [review]
This is looking good. Left some comments. IMHO this is not a draft anymore 😉

File: pkg/trace/api/debugger.go (hunk 6 above)
Comment (ojung) [reply pkg/trace/api/debugger.go:79]: Right, adding metrics is something I had in mind anyway but forgot. Thanks for the reminder 😅

File: pkg/trace/api/debugger.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+	// logsIntakeURLTemplate specifies the template for obtaining the profiling URL along with the site.
+	logsIntakeURLTemplate = "https://http-intake.logs.%s/v1/input"
+	// logsIntakeDefaultURL specifies the default intake API URL.
+	logsIntakeDefaultURL = "https://http-intake.logs.datadoghq.com/v1/input"
+
+	// debuggerRequestMetricsPrefix is the prefix for the following debugger metric names.
+	debuggerRequestMetricsPrefix       = "datadog.trace_agent.debugger."
+	debuggerRequestCountMetricsName    = debuggerRequestMetricsPrefix + "proxy_request"
+	debuggerRequestDurationMetricsName = debuggerRequestMetricsPrefix + "proxy_request_duration_ms"
+	debuggerRequestErrorMetricsName    = debuggerRequestMetricsPrefix + "proxy_request_error"
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:34]: We don't use constants for metrics anywhere in `pkg/trace`. I know you copied these from the AppSec package. It happens there because I gave them freedom to use whatever coding style they want since it's a different team. Writing this now I realise it's not the best practice and there should be repository-wide coding conventions, but it is what it is. Can you please just inline them as strings?

File: pkg/trace/api/debugger.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+)
+
+// debuggerProxyHandler returns an http.Handler proxying requests to the logs intake. If the logs intake url cannot be
+// parsed, the returned handler will always return http.StatusInternalServerError with a clarifying message.
+func (r *HTTPReceiver) debuggerProxyHandler() http.Handler {
+	tags := fmt.Sprintf("host:%s,default_env:%s,agent_version:%s", r.conf.Hostname, r.conf.DefaultEnv, info.Version)
+	orch := r.conf.FargateOrchestrator
+	if orch != fargate.Unknown {
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:42]:
```suggestion
	if orch := r.conf.FargateOrchestrator; orch != fargate.Unknown {
```
This is better since the var isn't really used anywhere else outside the `if` clause.

File: pkg/trace/api/debugger.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+type roundTripper struct {
+	url *url.URL
+	key string
+	rt  http.RoundTripper
+}
+
+func (r *roundTripper) RoundTrip(req *http.Request) (res *http.Response, err error) {
+	now := time.Now()
+	defer func() {
+		tags := []string{"path:" + req.URL.Path}
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:87]: I don't think this needs to be a tag. Unless you have multiple paths you'll be using today, in which case ignore my comment.

File: pkg/trace/api/debugger.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+		metrics.Count(debuggerRequestCountMetricsName, 1, tags, 1)
+		metrics.Timing(debuggerRequestDurationMetricsName, time.Since(now), tags, 1)
+		if err != nil {
+			tags = append(tags, fmt.Sprintf("error:%s", fmt.Sprintf("%T", err)))
+			metrics.Count(debuggerRequestErrorMetricsName, 1, tags, 1)
+		}
+	}()
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:94]:
```suggestion
	defer func(now) {
		metrics.Count("datadog.trace_agent.debugger.proxy_request", 1, nil, 1)
		metrics.Timing("datadog.trace_agent.debugger.proxy_request_duration_ms", time.Since(now), nil, 1)
		if err != nil {
			metrics.Count("datadog.trace_agent.debugger.proxy_request_error", 1, fmt.Sprintf("error:%s", fmt.Sprintf("%T", err)), 1)
		}
	}(time.Now())
```

File: pkg/trace/api/api.go [hunk 11]
```diff
@@ -762,3 +762,23 @@ func getMediaType(req *http.Request) string {
...
+			// that net/http gives it: Go-http-client/1.1
+			// See https://codereview.appspot.com/7532043
+			req.Header.Set("User-Agent", "")
+		}
+		containerID := req.Header.Get(headerContainerID)
+		if ctags := getContainerTags(containerID); ctags != "" {
+			req.Header.Set("X-Datadog-Container-Tags", ctags)
+		}
+		req.Header.Set("X-Datadog-Additional-Tags", tags)
+		metrics.Count("datadog.trace_agent."+metricSuffix, 1, nil, 1)
```
Comment (gbbr) [inline pkg/trace/api/api.go:782]: I think you can remove this along with the function argument. This is a duplicate of “proxy_request” metric. At this point we are barely setting headers so it’s unfair to say that any request happened. In profiling, you can move the metric to the RoundTrip method. 

I also think it’s slightly odd that this func is in a file where it’s not used but I understand the reasoning and don’t feel strongly about it.

File: pkg/trace/api/debugger.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+	defer func(now time.Time) {
+		var tags []string
+		metrics.Count("datadog.trace_agent.debugger.proxy_request", 1, tags, 1)
+		metrics.Timing("datadog.trace_agent.debugger.proxy_request_duration_ms", time.Since(now), tags, 1)
+		if err != nil {
+			tags := append(tags, fmt.Sprintf("error:%s", fmt.Sprintf("%T", err)))
+			metrics.Count("datadog.trace_agent.debugger.proxy_request_error", 1, tags, 1)
+		}
+	}(time.Now())
+	req.Header.Set("DD-API-KEY", r.key)
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:87]: Why not pass this as an argument to the director function and define all headers there? That’s the role of the director, not here in the RoundTripper, if possible. 

I realize this affects the “de-duplication plan” but it’s small differences like this that make me think that de-duplicating code here is not worth it because we are making compromises in terms of coupling and quality to obtain a result which isn’t that beneficial. 

I think we should either remove the helper and have everything inline or just add a function which takes an http.Header and adds container tags, user agent, etc.

File: pkg/trace/api/api.go (hunk 11 above)
Comment (ojung) [reply pkg/trace/api/api.go]: Yeah, I just noticed that as well. Pushed a fix already.

File: pkg/trace/api/debugger.go (hunk 12 above)
Comment (ojung) [reply pkg/trace/api/debugger.go:87]: Jup, I agree. I'll push an alternative in a sec.

File: pkg/trace/api/api.go [hunk 13]
```diff
@@ -762,3 +762,20 @@ func getMediaType(req *http.Request) string {
 	}
 	return mt
 }
+
+// setCommonHeaders enriches requests with headers setting custom container and additional tags and overwrites the
+// user agent header as well as sets the Via header.
+func setCommonHeaders(req *http.Request, tags string) {
```
Comment (gbbr) [inline pkg/trace/api/api.go:768]: This is better, but we don't need the request anymore.
```suggestion
func setCommonHeaders(h http.Header, tags string) {
```

File: pkg/trace/api/profiles.go [hunk 14]
```diff
@@ -2,8 +2,9 @@ package api
 
 import (
 	"bytes"
-	"context"
 	"fmt"
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
```
Comment (gbbr) [inline pkg/trace/api/profiles.go:6]: These imports are in the wrong group...

File: pkg/trace/api/profiles.go [hunk 15]
```diff
@@ -138,6 +117,9 @@ type multiTransport struct {
 }
 
 func (m *multiTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	defer func() {
+		metrics.Count("datadog.trace_agent.profile", 1, nil, 1)
+	}()
```
Comment (gbbr) [inline pkg/trace/api/profiles.go:122]: No need to `defer` this. It does not need to happen at the end.

File: releasenotes/notes/live-debugger-endpoint-9fa1ddaa9412bdfe.yaml [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+issues:
+  - |
+deprecations:
+  - |
+security:
+  - |
+fixes:
+  - |
+other:
+  - |
```
Comment (gbbr) [inline releasenotes/notes/live-debugger-endpoint-9fa1ddaa9412bdfe.yaml:25]: All the other sections need to be deleted.

File: pkg/trace/api/debugger.go [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+
+// debuggerProxyHandler returns an http.Handler proxying requests to the logs intake. If the logs intake url cannot be
+// parsed, the returned handler will always return http.StatusInternalServerError with a clarifying message.
+func (r *HTTPReceiver) debuggerProxyHandler() http.Handler {
+	tags := fmt.Sprintf("host:%s,default_env:%s,agent_version:%s", r.conf.Hostname, r.conf.DefaultEnv, info.Version)
+	if orch := r.conf.FargateOrchestrator; orch != fargate.Unknown {
+		tag := fmt.Sprintf("orchestrator:fargate_%s", strings.ToLower(string(orch)))
+		tags = tags + "," + tag
+	}
+	intake := logsIntakeDefaultURL
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:39]: I think we can drop `logsIntakeDefaultURL` and just default this value to:
```suggestion
	intake := fmt.Sprintf(logsIntakeURLTemplate, config.DefaultSite)
```
What do you think?

File: pkg/trace/api/debugger.go [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+	if v := config.Datadog.GetString("apm_config.debugger_dd_url"); v != "" {
+		intake = v
+	} else if site := config.Datadog.GetString("site"); site != "" {
+		intake = fmt.Sprintf(logsIntakeURLTemplate, site)
+	}
+	target, err := url.Parse(intake)
+	if err != nil {
+		return debuggerErrorHandler(fmt.Errorf("error parsing snapshot intake URL %s: %v", intake, err))
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:47]:
```suggestion
        log.Errorf("Live Debugger proxy is OFF due to invalid debugger_dd_url: %v", err)
		return debuggerErrorHandler(fmt.Errorf("error parsing snapshot intake URL %s: %v", intake, err))
```
We shouldn't wait for someone to make a first request before knowing that there's a critical issue here. This might even have to be `log.Criticalf` since we're not technically running if we hit this branch.

File: pkg/trace/config/config.go [hunk 19]
```diff
@@ -141,16 +141,16 @@ type Tag struct {
 // New returns a configuration with the default values.
 func New() *AgentConfig {
 	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
-	isFargate := fargate.IsFargateInstance(ctx)
+	orch := fargate.GetOrchestrator(ctx)
 	cancel()
 	if err := ctx.Err(); err != nil && err != context.Canceled {
-		log.Errorf("Error detecting Fargate instance: %v. Assuming not. If you are in a Fargate instance, this will cause problems.", err)
+		log.Warnf("Failed to get Fargate orchestrator. This may cause issues if you are in a Fargate instance: %v", err)
```
Comment (gbbr) [inline pkg/trace/config/config.go:147]: I think this should be more than just a warning since it can have some serious implications on APM stats and billing. Why did you change it?

**Review by gbbr — COMMENTED** [review]
This is fine by me, despite some small nits.

File: pkg/trace/api/api.go [hunk 20]
```diff
@@ -762,3 +762,20 @@ func getMediaType(req *http.Request) string {
 	}
 	return mt
 }
+
+// setCommonHeaders enriches requests with headers setting custom container and additional tags and overwrites the
+// user agent header as well as sets the Via header.
+func setCommonHeaders(h *http.Header, tags string) {
```
[the remark this answers is not in the document]
Comment (ojung) [reply pkg/trace/api/api.go:768]: Ah, cool! TIL :)

File: pkg/trace/config/config.go (hunk 19 above)
Comment (ojung) [reply pkg/trace/config/config.go:147]: It was warning in the other place. I didn't intend to change it.

File: pkg/trace/api/profiles.go [hunk 21]
```diff
@@ -101,25 +94,11 @@ func errorHandler(err error) http.Handler {
...
-		if ctags := getContainerTags(containerID); ctags != "" {
-			req.Header.Set("X-Datadog-Container-Tags", ctags)
-		}
-		req.Header.Set("X-Datadog-Additional-Tags", tags)
-		metrics.Count("datadog.trace_agent.profile", 1, nil, 1)
-		// URL, Host and key are set in the transport for each outbound request
-	}
 	logger := logutil.NewThrottled(5, 10*time.Second) // limit to 5 messages every 10 seconds
+	// URL, Host and key are set in the transport for each outbound request
+	// headers are set in the director
```
Comment (gbbr) [inline pkg/trace/api/profiles.go:99]: No they're not 😉 We can get rid of this comment now.

File: pkg/trace/api/profiles.go [hunk 22]
```diff
@@ -101,25 +94,11 @@ func errorHandler(err error) http.Handler {
...
 	return &httputil.ReverseProxy{
-		Director:  director,
+		Director:  func(req *http.Request) { setCommonHeaders(req.Header, tags) },
```
Comment (gbbr) [inline pkg/trace/api/profiles.go:101]:
```suggestion
		Director:  func(req *http.Request) {
				setCommonHeaders(req.Header, tags)
				// URL, Host and API Keys are set by the Transport implementation multiTransport.
		},
```
And we can remove the comment above.

File: pkg/trace/api/debugger.go [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+		msg := fmt.Sprintf("Debugger Proxy is OFF: %v", err)
+		http.Error(w, msg, http.StatusInternalServerError)
+	})
+}
+
+// newDebuggerProxy returns a new httputil.ReverseProxy proxying and augmenting requests with headers containing the tags.
+func newDebuggerProxy(rt http.RoundTripper, target *url.URL, key string, tags string) *httputil.ReverseProxy {
+	logger := logutil.NewThrottled(5, 10*time.Second) // limit to 5 messages every 10 seconds
+	director := func(req *http.Request) {
+		newTags := tags
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:65]: I think we don't need a new variable here, and can just continue using `tags` (it's already a copy of the original string if that was your concern).

File: pkg/trace/api/debugger.go [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+	logsIntakeURLTemplate = "https://http-intake.logs.%s/v1/input"
+)
+
+// debuggerProxyHandler returns an http.Handler proxying requests to the logs intake. If the logs intake url cannot be
+// parsed, the returned handler will always return http.StatusInternalServerError with a clarifying message.
+func (r *HTTPReceiver) debuggerProxyHandler() http.Handler {
+	tags := fmt.Sprintf("host:%s,default_env:%s,agent_version:%s", r.conf.Hostname, r.conf.DefaultEnv, info.Version)
+	if orch := r.conf.FargateOrchestrator; orch != fargate.Unknown {
+		tag := fmt.Sprintf("orchestrator:fargate_%s", strings.ToLower(string(orch)))
+		tags = tags + "," + tag
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:36]:
```suggestion
		tags = tags + ",orchestrator:fargate_" + strings.ToLower(string(orch))
```

File: pkg/trace/api/debugger.go [hunk 25]
```diff
@@ -0,0 +1,104 @@
...
+	}
+	intake := fmt.Sprintf(logsIntakeURLTemplate, config.DefaultSite)
+	if v := config.Datadog.GetString("apm_config.debugger_dd_url"); v != "" {
+		intake = v
+	} else if site := config.Datadog.GetString("site"); site != "" {
+		intake = fmt.Sprintf(logsIntakeURLTemplate, site)
+	}
+	target, err := url.Parse(intake)
+	if err != nil {
+		msg := fmt.Errorf("error parsing snapshot intake URL %s: %v", intake, err)
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:46]: I think you might want to separate the two here. The log message should be more descriptive (and also because log messages need to start with a capital letter, but not errors).

So we should log something like:
```go
log.Criticalf("Error parsing Live Debugger intake URL %q: %v", intake, err)
```
And then use your current message for the handler. (`%q` will add quotes around the string)

File: pkg/trace/api/debugger.go [hunk 26]
```diff
@@ -0,0 +1,104 @@
...
+func newDebuggerProxy(rt http.RoundTripper, target *url.URL, key string, tags string) *httputil.ReverseProxy {
+	logger := logutil.NewThrottled(5, 10*time.Second) // limit to 5 messages every 10 seconds
+	director := func(req *http.Request) {
+		newTags := tags
+		containerID := req.Header.Get(headerContainerID)
+		if ct := getContainerTags(containerID); ct != "" {
+			newTags = fmt.Sprintf("%s,%s", newTags, ct)
+		}
+		q := req.URL.Query()
+		if oldTags := q.Get("ddtags"); oldTags != "" {
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:71]: Is this in case the client also adds tags in the query string? I think a better name is `qtags` (it relates to your other variable `q` which is the query string). But if you really want `oldTags` don't mind me :) It's just that their not "old", they are just the additional query tags sent by the client.

File: pkg/trace/api/debugger.go [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+			newTags = fmt.Sprintf("%s,%s", newTags, oldTags)
+		}
+		q.Set("ddtags", newTags)
+		newTarget := target
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:75]: This is creating a copy of the pointer, not of the value. You should do this instead:
```go
targetcp := *target
```

File: pkg/trace/api/debugger.go [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+		newTarget.RawQuery = q.Encode()
+		req.Header.Set("DD-API-KEY", key)
+		req.URL = newTarget
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:78]: And then here...
```suggestion
		req.URL = &targetcp
```

File: pkg/trace/api/debugger.go (hunk 23 above)
Comment (ojung) [reply pkg/trace/api/debugger.go:65]: But it will live longer than each invocation of the director function 😅

Comment (gbbr) [reply pkg/trace/api/debugger.go:65]: Ah yes, you're right. My bad. Carry on. Maybe we can call it `ddtags`? It will match the query string. I'm not a very big fan of camelCaseVariables that are short lived. I notice in Go they are usually very short if they only last a few lines. Would you mind?

Comment (ojung) [reply pkg/trace/api/debugger.go:65]: Not at all :+1:

File: pkg/trace/api/debugger.go [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+	}
+	intake := fmt.Sprintf(logsIntakeURLTemplate, config.DefaultSite)
+	if v := config.Datadog.GetString("apm_config.debugger_dd_url"); v != "" {
+		intake = v
+	} else if site := config.Datadog.GetString("site"); site != "" {
+		intake = fmt.Sprintf(logsIntakeURLTemplate, site)
+	}
+	target, err := url.Parse(intake)
+	if err != nil {
+		log.Criticalf("Error parsing snapshot intake URL %q: %v", intake, err)
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:45]:
```suggestion
		log.Criticalf("Error parsing live debugger intake URL %q: %v", intake, err)
```
"snapshot" is too ambiguous I think, considering that the trace-agent now handles multiple products...

The below error (in the error handler) is fine, that's within context.

File: pkg/trace/api/debugger.go [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+		}
+		q := req.URL.Query()
+		if qtags := q.Get("ddtags"); qtags != "" {
+			ddtags = fmt.Sprintf("%s,%s", ddtags, qtags)
+		}
+		q.Set("ddtags", ddtags)
+		newTarget := *target
+		newTarget.RawQuery = q.Encode()
+		req.Header.Set("DD-API-KEY", key)
+		log.Infof("newTarget: %v", newTarget)
```
Comment (gbbr) [inline pkg/trace/api/debugger.go:76]: I think we can now get rid of this.

File: pkg/trace/api/debugger_test.go [hunk 31]
```diff
@@ -0,0 +1,137 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
Comment (gbbr) [inline pkg/trace/api/debugger_test.go:1]: I realise you've copied these tests from the other proxy. That's fine. But I was wondering whether there would be a way to easily de-dupe them by perhaps creating a function which runs the tests and takes as input a config and an `http.ReverseProxy` ?

If it's complicated or not an easy approach, it's fine to leave it as it is.

**Review by gbbr — COMMENTED** [review]
I think we're good. Let's make sure the client PR also gets merged, and that we test it thoroughly with it.

File: pkg/trace/api/debugger_test.go (hunk 31 above)
Comment (ojung) [reply pkg/trace/api/debugger_test.go:1]: Hm, I'm not sure it's worth the effort or feasible at all. The tests only resemble each other in structure. But they differ in the tested types and in the assertions.

File: pkg/trace/api/debugger_test.go [hunk 32]
```diff
@@ -0,0 +1,158 @@
...
+	slurp, err := ioutil.ReadAll(rec.Result().Body)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if string(slurp) != "OK" {
+		t.Fatal("did not proxy")
+	}
+}
+
+func testHandler(conf *traceconfig.AgentConfig, assertions []func(req *http.Request) error) (string, error) {
```
Comment (gbbr) [inline pkg/trace/api/debugger_test.go:62]: I think you were supposed to delete this :)

Comment (ojung) [reply pkg/trace/api/debugger_test.go:62]: Ah, yep.

**Review by gbbr — APPROVED** [review]
